File: app/src/main/java/com/example/smartspend/viewmodel/HistoryViewModel.java
// app/src/main/java/com/example/smartspend/viewmodel/HistoryViewModel.java
package com.example.smartspend.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.example.smartspend.database.AppDatabase;
import com.example.smartspend.database.TransactionDao;
import com.example.smartspend.model.Transaction;

import java.util.List;

public class HistoryViewModel extends AndroidViewModel {

    private final TransactionDao dao;
    private final int userId;

    // Paramètres de filtrage observables
    private final MutableLiveData<long[]> dateRange = new MutableLiveData<>();
    private final MutableLiveData<String> searchQuery = new MutableLiveData<>();

    // LiveData final exposé à l'Activity
    public final LiveData<List<Transaction>> transactions;

    public HistoryViewModel(@NonNull Application application, int userId) {
        super(application);
        this.userId = userId;
        dao = AppDatabase.getInstance(application).transactionDao();

        // Transformation : selon l'état des filtres, choisir la bonne requête
        transactions = Transformations.switchMap(dateRange, range -> {
            if (range == null) {
                String q = searchQuery.getValue();
                if (q != null && !q.isEmpty())
                    return dao.searchTransactions(userId, q);
                return dao.getAllTransactions(userId);
            }
            return dao.getTransactionsByDateRange(userId, range[0], range[1]);
        });
    }

    public void filterByDateRange(long start, long end) {
        dateRange.setValue(new long[]{start, end});
    }

    public void resetFilter() {
        dateRange.setValue(null);
    }

    public void search(String query) {
        searchQuery.setValue(query);
        dateRange.setValue(null); // reset filtre date si on cherche
    }
}

File: app/src/main/java/viewmodel/HistoryViewModel.kt
package com.team.smartspend.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.switchMap
import com.team.smartspend.database.AppDatabase
import com.team.smartspend.model.Transaction

/**
 * HistoryViewModel — alimente l'écran Historique (Membre 4, adapté).
 *
 * Selon l'état des filtres, il choisit automatiquement la bonne requête Room :
 *   - une plage de dates active     -> getTransactionsByDate
 *   - une recherche active          -> searchTransactions
 *   - aucun filtre                  -> getAllTransactions
 *
 * Le tout est réactif : la liste se met à jour en temps réel via LiveData.
 */
class HistoryViewModel(application: Application) : AndroidViewModel(application) {

    private val dao = AppDatabase.getDatabase(application).transactionDao()

    // État des filtres (déclencheurs)
    private val dateRange = MutableLiveData<Pair<Long, Long>?>(null)
    private val searchQuery = MutableLiveData<String>("")

    // Liste finale exposée à l'écran : recalculée dès qu'un filtre change.
    val transactions: LiveData<List<Transaction>> = dateRange.switchMap { range ->
        if (range != null) {
            // Filtre par plage de dates prioritaire
            dao.getTransactionsByDate(range.first, range.second)
        } else {
            val q = searchQuery.value ?: ""
            if (q.isNotEmpty()) dao.searchTransactions(q)
            else dao.getAllTransactions()
        }
    }

    /** Filtre par plage de dates (date début / date fin). */
    fun filterByDateRange(start: Long, end: Long) {
        dateRange.value = Pair(start, end)
    }

    /** Réinitialise tous les filtres (réaffiche tout). */
    fun resetFilter() {
        searchQuery.value = ""
        dateRange.value = null
    }

    /** Recherche en temps réel (sur description + catégorie). */
    fun search(query: String) {
        searchQuery.value = query
        // Une recherche annule le filtre de date ; comme la liste dépend de
        // dateRange, on lui réassigne null pour forcer le recalcul.
        dateRange.value = null
    }
}
